core/src/header_chain_prover.rs

```diff
@@ -0,0 +1,761 @@
+pub enum HeaderChainProverError {
+    #[error("Error while de/serializing object")]
+    ProverDeSerializationError,
+    #[error("Wait for candidate batch to be ready")]
+    BatchNotReady,
+
+    #[error(transparent)]
+    Other(#[from] eyre::Report),
+}
+
+#[derive(Debug, Clone)]
+pub struct HeaderChainProver {
+    db: Database,
+    network: bitcoin::Network,
+    batch_size: u64,
+}
+
+impl HeaderChainProver {
+    /// Creates a new [`HeaderChainProver`] instance. Also saves a proof
+    /// assumption if specified in the config.
+    pub async fn new(
+        config: &BridgeConfig,
+        rpc: ExtendedRpc,
+    ) -> Result<Self, HeaderChainProverError> {
+        let db = Database::new(config).await.map_to_eyre()?;
+
+        if let Some(proof_file) = &config.header_chain_proof_path {
+            tracing::info!("Starting prover with assumption file {:?}.", proof_file);
+            let file = File::open(proof_file)
+                .wrap_err_with(|| format!("Failed to open proof assumption file {proof_file:?}"))?;
+
+            let mut reader = BufReader::new(file);
+            let mut assumption = Vec::new();
+            reader
+                .read_to_end(&mut assumption)
+                .wrap_err("Can't read assumption file")?;
+
+            let proof: Receipt = borsh::from_slice(&assumption)
+                .wrap_err(HeaderChainProverError::ProverDeSerializationError)?;
+            let proof_output: BlockHeaderCircuitOutput = borsh::from_slice(&proof.journal.bytes)
+                .wrap_err(HeaderChainProverError::ProverDeSerializationError)?;
+
+            // Create block entry, if not exists.
+            let block_hash = BlockHash::from_raw_hash(
+                Hash::from_slice(&proof_output.chain_state.best_block_hash).map_to_eyre()?,
+            );
+            let block_header = rpc
+                .client
+                .get_block_header(&block_hash)
+                .await
+                .wrap_err("Failed to get block header")?;
+            let block_height = rpc
+                .client
+                .get_block_info(&block_hash)
+                .await
+                .map(|info| info.height)
+                .wrap_err("Failed to get block info")?;
+            tracing::info!(
+                "Adding proof assumption for a block with hash of {:?}, header of {:?} and height of {}",
+                block_hash,
+                block_header,
+                block_height
+            );
+
+            // If an unproven block in database already exists, it shouldn't
+            // effect anything.
+            // PS: This also ignores other db errors but there are other places
+            // where we check for those errors.
+            let _ = db
+                .save_unproven_finalized_block(
+                    None,
+                    block_hash,
+                    block_header,
+                    proof_output.chain_state.block_height.into(),
+                )
+                .await;
```

Comment:
I'm against this ignoring. At the very least, we have to do an inspect_err with tracing::error

```suggestion
            let _ = db
                .save_unproven_finalized_block(
                    None,
                    block_hash,
                    block_header,
                    proof_output.chain_state.block_height.into(),
                )
                .await.inspect_err(|e| tracing::error!("Failed to save unproven finalized block to db {e:?}"));
```

Reply:
https://github.com/chainwayxyz/clementine/pull/653/commits/a6c21398ea2137ac7938d1c52ffbec099398d426

With the aggregator distributing initial header chain proof PR, this call will be removed. Because of this, I don't think it is worth to implement a proper error check here.